Fix synchronization issues in radix sort and histogram

dmitriy-sobolev opened the pull request:
Value (`__buf_val`) and Count (`__buf_count`) buffers store data in either local or global memory. Let's use group barriers with proper fences (local or global memory fences) to avoid memory contention issues, which are observed on Xe2 architectures.

The fix changes how `__group_barrier` is defined, so the PR additionally includes what is done in #1988 to avoid conflicts and fix another issue related to the barriers: we use SYCL 1.2.1 barriers, but oneDPL claims to be SYCL 2020 compatible.

SergeyKopienko commented:
Taking into account the discussion from https://github.com/uxlfoundation/oneDPL/issues/1679
I think that probably the switch which barrier exactly we should use in all oneAPI code - may be incorrect approach...

adamfidel commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h:
```diff
@@ -173,10 +173,23 @@ struct __subgroup_radix_sort
+                            else
+                                __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
+                        };
+
```
This just a thought, but could we use a single lambda here and pass the boolean constant as a parameter to it? Something like:
```c++
    auto __mem_adjusted_barrier = [__it](auto __is_slm) {
        if constexpr (decltype(__is_slm)::value)
            __dpl_sycl::__group_barrier(__it);
        else
            __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
    };
```
And when used:
```c++
   __mem_adjusted_barrier(_SLM_tag_val{});
```
```c++   
  __mem_adjusted_barrier(_SLM_counter{});
```

See https://godbolt.org/z/GMx64dTb5.

MikeDvorskiy replied:
Why do we need a compile time tag here?
Whether is it simpler to call with a constant, wrapped into `__dpl_sycl::name_xx`, like?

`__dpl_sycl::__group_barrier(__it, __dpl_sycl::fence_global_and_local);`

dmitriy-sobolev replied:
@MikeDvorskiy, I do not understand how you suggesting is supposed to work. We need to use different fence arguments. Could you be more specific? Perhaps I just do not get your idea.

dmitriy-sobolev replied:
Meanwhile, I've implemented what Adam suggested.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h:
```diff
@@ -173,10 +173,23 @@ struct __subgroup_radix_sort
+                            if constexpr (_SLM_counter::value)
+                                __dpl_sycl::__group_barrier(__it);
+                            else
+                                __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
```
Since `__counters` is entirely through global memory in this case, is `__dpl_sycl::__fence_space_global` sufficient here?

dmitriy-sobolev replied:
Indeed, `__fence_space_global` is sufficient. I've applied your suggestion.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -225,17 +226,45 @@ __get_accessor_size(const _Accessor& __accessor)
+
+#if ONEDPL_SYCL121_GROUP_BARRIER
+template <sycl::access::fence_space _Space>
+struct __fence_space
```
I would suggest writing 3floor C++ code....
Actually? we already have the simple approach for wrapping sycl constants. For example,
see `__target_device` usage.

dmitriy-sobolev replied:
The problem is that I need to provide some no-op alternative. Defining structures helps to achieve it. Could you provide an example in this context?

danhoeflinger replied:
Perhaps you can do this a little simpler...
```
#if ONEDPL_SYCL121_GROUP_BARRIER
constexpr sycl::access::fence_space __fence_space_local = sycl::access::fence_space::local_space;
constexpr sycl::access::fence_space __fence_space_global = sycl::access::fence_space::global_space;
constexpr sycl::access::fence_space __fence_space_global_and_local = sycl::access::fence_space::global_and_local;
#else
constexpr int __fence_space_local = 0;
constexpr int __fence_space_global = 0;
constexpr int __fence_space_global_and_local = 0;
#endif // ONEDPL_SYCL121_GROUP_BARRIER


template <typename _Item, typename _Space = decltype(__fence_space_local)>
void
__group_barrier(_Item __item, [[maybe_unused]] _Space __space = __fence_space_local)
{
#if ONEDPL_SYCL121_GROUP_BARRIER
    __item.barrier(__space);
#elif _ONEDPL_SYCL2020_GROUP_BARRIER_PRESENT
    sycl::group_barrier(__item.get_group(), sycl::memory_scope::work_group);
#else
#    error "sycl::group_barrier is not supported, and no alternative is available"

#endif
}
```
Does this work?  maybe there is a better "dummy" type / value than int, but otherwise I think it would work.  

You would just need to remove the `{}` from `__dpl_sycl::__fence_space_global{}` above I think.

danhoeflinger replied:
I guess if we are following the example of `__target_device` we can do a similar trick to create a typename `__fence_space_t` which is either `sycl::access::fence_space` or `int` depending on `ONEDPL_SYCL121_GROUP_BARRIER`.

dmitriy-sobolev replied:
The option with global constants looks good. I've implemented it with adding a dummy type.

SergeyKopienko commented:
My assumptions that the changes from this PR is a little bit overcomplicated.
Let me propose alternative approach for this fix: https://github.com/uxlfoundation/oneDPL/pull/2055

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -225,17 +226,45 @@ __get_accessor_size(const _Accessor& __accessor)
     sycl::group_barrier(__item.get_group(), sycl::memory_scope::work_group);
 #else
-    __item.barrier(sycl::access::fence_space::local_space);
+#    error "sycl::group_barrier is not supported, and no alternative is available"
```
Whether do we get potential issues here with user's code in case when neither `ONEDPL_SYCL121_GROUP_BARRIER `neither `_ONEDPL_SYCL2020_GROUP_BARRIER_PRESENT `is defined?

SergeyKopienko commented:
About
"The fix changes how __group_barrier is defined, so the PR additionally includes what is done in https://github.com/uxlfoundation/oneDPL/pull/1988 to avoid conflicts and fix another issue related to the barriers: we use SYCL 1.2.1 barriers, but oneDPL claims to be SYCL 2020 compatible."
- I believe it's incorrect moment to change that before release. Let's do this change later in the next release.

dmitriy-sobolev commented:
> Taking into account the discussion from #1679 I think that probably the switcher which barrier exactly we should use in all `oneDPL` code - may be incorrect approach...

I briefly checked the other places: they all seem to be using local memory. But let me delve deeper.

dmitriy-sobolev commented:
> About "The fix changes how __group_barrier is defined, so the PR additionally includes what is done in #1988 to avoid conflicts and fix another issue related to the barriers: we use SYCL 1.2.1 barriers, but oneDPL claims to be SYCL 2020 compatible."
> 
> * I believe it's incorrect moment to change that before release. Let's do this change later in the next release.

Well, technically, the changes and fixes should be as isolated as possible, especially when they are close to a release.
But, when I implemented the fix with barriers, keeping in mind that it should not conflict with #1988, I implemented the major part of #1988, which had already been approved. Why not to fix that as well... Anyway, the alternative version you provided is helpful if others think the opposite.

mmichel11 commented:
> > About "The fix changes how __group_barrier is defined, so the PR additionally includes what is done in #1988 to avoid conflicts and fix another issue related to the barriers: we use SYCL 1.2.1 barriers, but oneDPL claims to be SYCL 2020 compatible."
> > 
> > * I believe it's incorrect moment to change that before release. Let's do this change later in the next release.
> 
> Well, technically, the changes and fixes should be as isolated as possible, especially when they are done close to a release. But, when I implemented the fix with barriers, keeping in mind that it should not conflict with #1988, I implemented the major part of #1988, which had already been approved. Why not to fix that as well... Anyway, the alternative version you provided is helpful if others think the opposite.

My opinion is the approach here is fine for the release. Since `ONEDPL_SYCL121_GROUP_BARRIER` defaults to 1 for the Intel SYCL implementation, we still get the same SYCL 1.2.1 barrier by default. Only if the user sets the variable themselves or a separate SYCL compiler is used, we get `sycl::group_barrier`. Using `sycl::group_barrier` in the case of a separate SYCL compiler is likely much safer than what we do now which is assume it supports the SYCL 1.2.1 barrier.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -225,17 +226,45 @@ __get_accessor_size(const _Accessor& __accessor)
+#    if _ONEDPL_LIBSYCL_VERSION
+#        define ONEDPL_SYCL121_GROUP_BARRIER 1
+#    else
+#        define ONEDPL_SYCL121_GROUP_BARRIER 0
```
This is the only think I see as "risky" in the PR now that we are past code freeze.  

This changes the default behavior for non-intel LLVM compilers to assume sycl 2020 group barrier support, where previous to this PR, we were assuming sycl 1.2.1 support.

Would it make sense to have a `SYCL_LANGUAGE_VERSION >= 202000` check here and decide for non-intel compilers that way?

(This remains a change which may not be "required" at a time where we should resist making changes).

danhoeflinger replied:
I think we have the following decision to make for the else branch here:
1) Minimize change post code freeze
`#define ONEDPL_SYCL121_GROUP_BARRIER 1`

2) Assume sycl 2020 for non intel llvm compiler 
`#define ONEDPL_SYCL121_GROUP_BARRIER 0`

3) Check sycl spec version macro and decide based on that
`#define ONEDPL_SYCL121_GROUP_BARRIER SYCL_LANGUAGE_VERSION >= 202000`


I don't have as good a grasp on what should be expected for non intel-llvm sycl compilers, so I'm interested what others think here.

dmitriy-sobolev replied:
I suggest making a decision on Monday as it is late. In the meantime, I will test the PR using icx and acpp.

My personal preference is option 2: oneDPL declares SYCL 2020 compatibility, so this change can be treated as a bug fix, while there is a backup option to manually set `ONEDPL_SYCL121_GROUP_BARRIER` for those who rely on these barriers.

danhoeflinger replied:
Does acpp correctly define `SYCL_LANGUAGE_VERSION`?

If so, would (3) be a more sycl specification compliant version of basically the same as (2)?

dmitriy-sobolev replied:
> Does acpp correctly define SYCL_LANGUAGE_VERSION?

It does: https://github.com/AdaptiveCpp/AdaptiveCpp/blob/e1c6c6ede4c17877a2005f876afc869f11fba5be/include/hipSYCL/sycl/sycl.hpp#L27C1-L27C38

> If so, would (3) be a more sycl specification compliant version of basically the same as (2) at least for acpp?

Agree (3) looks better than (2) with this additional check. The check should be slightly different, though. If it is SYCL 2020, then define `ONEDPL_SYCL121_GROUP_BARRIER` to `0`.

danhoeflinger replied:
yes you are correct.  Thanks for the correction.

dmitriy-sobolev replied:
I checked that ACPP supports SYCL 2020 barriers. I am now slightly inclined towards keeping option (2) as we want to have only SYCL 2020 barriers in the future, and we do not now other implementations still having only SYCL 1.2.1 barriers, but which are SYCL 2020 compliant in other places.

danhoeflinger replied:
I'm OK with that option, but lets leave a comment explaining it.

dmitriy-sobolev replied:
Added:
> // For safety, assume that other SYCL implementations comply with SYCL 2020, which is a oneDPL requirement.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -241,26 +241,22 @@ __get_accessor_size(const _Accessor& __accessor)
+inline constexpr sycl::access::fence_space __fence_space_global = sycl::access::fence_space::global_space;
+inline constexpr sycl::access::fence_space __fence_space_global_and_local = sycl::access::fence_space::global_and_local;
+#else
+struct __fence_space_dummy{}; // No-op dummy type since SYCL 2020 does specify memory fence spaces in group barriers
```
```suggestion
struct __fence_space_dummy{}; // No-op dummy type since SYCL 2020 does not specify memory fence spaces in group barriers
```

SergeyKopienko commented:
If somebody will have interest for https://github.com/uxlfoundation/oneDPL/pull/2055, please reopen it.

danhoeflinger commented:
> I briefly checked the other places: they all seem to be using local memory. 
>But let me delve deeper.

It looks like one implementation of histogram may have a similar issue (my fault):

`__clear_wglocal_histograms` when called from `__histogram_general_private_global_atomics_submitter` is operating on global data. 

I can come up with a fix if you want.

danhoeflinger commented:
I've added https://github.com/uxlfoundation/oneDPL/pull/2056 in case we want to include it here.  The rest of the usages I believe should be in the local memory space.  If we have any question here with the changes, lets wait because this implementation requires a very large number of bins to be used.

SergeyKopienko commented:
I think as far as @dmitriy-sobolev introduced three new types here, probably required to declare that they are device-copyable:
- `__fence_space_local`
- '__fence_space_global'
- `__fence_space_global_and_local'

SergeyKopienko commented:
Additional question: in the current state of this PR the value `__fence_space_global_and_local` is never used.
So should we really declare this value `__fence_space_global_and_local` ?

dmitriy-sobolev commented:
> Additional question: in the current state of this PR the value `__fence_space_global_and_local` is never used. So should we really declare this value `__fence_space_global_and_local` ?

They were indeed unnecessary, I've removed them.

dmitriy-sobolev commented:
> I think as far as @dmitriy-sobolev introduced three new types here, probably required to declare that they are device-copyable (`is_device_copyable`):
> 
> * `__fence_space_local`
> * `__fence_space_global`
> * `__fence_space_global_and_local`

I see a couple of problems here:
 - We should not copy these object onto a device.
 - `sycl::access::fence_space` is a trivially copyable type (enum), so should also be device copyable.

We discussed it offline, and the issue was work-arounded by declaring `get_fence` as static (a function this issue occurred), but I still have no understanding why the lack of `static` triggered that issue...

danhoeflinger reviewed (dismissed):
I agree with the first clang formatting complaint, not the others.  Otherwise, LGTM.

Probably makes sense to move the TODO above the line.

mmichel11 reviewed (approved):
LGTM
